pyupgrade/_plugins/typing_pep604.py

```diff
@@ -144,14 +144,18 @@ def visit_Subscript(
     if not _supported_version(state):
         return
 
-    # prevent rewriting forward annotations
+    # prevent rewriting forward annotations (unless they will be dequoted)
+    has_forward_annotation = (
+        (sys.version_info >= (3, 9) and _any_arg_is_str(node.slice)) or
+        (
+            sys.version_info < (3, 9) and
+            isinstance(node.slice, ast.Index) and
+            _any_arg_is_str(node.slice.value)
+        )
+    )
     if (
-            (sys.version_info >= (3, 9) and _any_arg_is_str(node.slice)) or
-            (
-                sys.version_info < (3, 9) and
-                isinstance(node.slice, ast.Index) and
-                _any_arg_is_str(node.slice.value)
-            )
+        has_forward_annotation and
+        'annotations' not in state.from_imports['__future__']
```

Comment:
while it probably doesn't actually matter in practice -- the `'annotations'` in check is very cheap and should probably come before the computation for forward annotations

Reply:
Good point, fixed. I've kept the `has_forward_annotation` variable for readability, but that is subjective. Let me know if you want me to make it a single big if statement. :)